### PR TITLE
Can combine D and C++ in a single target.

### PR DESCRIPTION
**Pull request by jpakkane**
This will _fail_ in CI with a segfault. The GDC runtime crashes when setting up the gc or something. I could not make it work even by running the commands manually. Reference page on D website [seems to be this one](https://dlang.org/spec/cpp_interface.html).

This should also make it simple to fix combining C++ and Fortran, which is #2685. Possibly even in this same PR once the D issue is fixed.

### CODE REVIEW

File: mesonbuild/build.py [hunk 1]
```diff
@@ -1015,12 +1017,19 @@ def get_clike_dynamic_linker(self):
...
+                    linker = all_compilers[l]
                 except KeyError:
                     raise MesonException(
                         'Could not get a dynamic linker for build target {!r}. '
                         'Requires a linker for language "{}", but that is not '
                         'a project language.'.format(self.name, l))
+                stdlib_args = []
+                added_languages = set()
+                for l in itertools.chain(self.compilers, dep_langs):
+                    if l != linker.language:
```
Comment (ignatenkobrain) [inline mesonbuild/build.py]: __[Flake8]__

[E741] ambiguous variable name 'l'

**Comment by jpakkane** [comment]
And it failed exactly as expected. @ximion, do you have any idea how to fix that?

**Comment by ximion** [comment]
@jpakkane I just tested this here: The test works perfectly with GDC >= 8.x and LDC >= 1.8, but fails with GDC 7.3.
So, the absolute best thing to do here would likely be to upgrade the CI to use GDC 8.x
(can currently be done by installing the gdc-8 package in Debian/Ubuntu and setting the DC environment variable to "gdc-8". GCC 8.x should also soonish be the default compiler in Debian Testing).

**Comment by ximion** [comment]
@jpakkane I just tested this here: The test works perfectly with GDC >= 8.x and LDC >= 1.8, but fails with GDC 7.3.
So, the absolute best thing to do here would likely be to upgrade the CI to use GDC 8.x
(can currently be done by installing the gdc-8 package in Debian/Ubuntu and setting the DC environment variable to "gdc-8". GCC 8.x should also soonish be the default compiler in Debian Testing).
CC: @ibuclaw

**Comment by jpakkane** [comment]
The CI image is Ubuntu. Bionic will be out soon and we'll update to that once it has stabilised a bit.

**Comment by ximion** [comment]
Okay - so if you update to Bionic, this issue will fix itself :-)
Alternatively, you could also compile this test with LDC. I didn't investigate this much, but it very much looks like a bug in the older compiler to me. All GDC versions << 8.x never yielded great results for me - the 8.x series is pretty good though (since it finally supports a recent version of the D specification and standard library, it is also much more usable in the real world now).

**Comment by ibuclaw** [comment]
@ximion - I guess part of your observation though is due to language spec aging pretty quickly (gdc-7 was at a pretty similar age gap to dmd as gdc-8 is now).  Give it a year and you'll say the same all over again and ask everyone to upgrade to gdc-9. :-)

**Comment by jpakkane** [comment]
If you use Clang as the C++ compiler then it fails with this in cppd:

```
#0  0x00005555555a6228 in gc.gc.Gcx.tryAlloc (this=this@entry=0x7fffffffdda0) at ../../../../src/libphobos/libdruntime/gc/gc.d:1736
#1  0x00005555555ade6b in gc.gc.Gcx.smallAlloc(ubyte, ref ulong, uint) (this=..., bin=<optimized out>, alloc_size=@0x7fffffffdde8: 128, bits=bits@entry=2)
    at ../../../../src/libphobos/libdruntime/gc/gc.d:1746
#2  0x00005555555ae316 in gc.gc.Gcx.alloc(ulong, ref ulong, uint) (bits=2, alloc_size=@0x7fffffffdde8: 128, size=72, this=...) at ../../../../src/libphobos/libdruntime/gc/gc.d:1725
#3  gc.gc.GC.mallocNoSync(ulong, uint, ref ulong, const(TypeInfo)) (ti=<optimized out>, alloc_size=@0x7fffffffdde8: 128, bits=2, size=72, this=...)
    at ../../../../src/libphobos/libdruntime/gc/gc.d:511
#4  gc.gc.GC.runLocked!(gc.gc.GC.mallocNoSync(ulong, uint, ref ulong, const(TypeInfo)), gc.gc.mallocTime, gc.gc.numMallocs, ulong, uint, ulong, const(TypeInfo)).runLocked(ref ulong, ref uint, ref ulong, ref const(TypeInfo)) (_param_3=<synthetic pointer>: <optimized out>, _param_2=@0x7fffffffdde8: 128, _param_1=<synthetic pointer>: <optimized out>, 
    _param_0=<synthetic pointer>: <optimized out>, this=...) at ../../../../src/libphobos/libdruntime/gc/gc.d:371
#5  gc.gc.GC.malloc(ulong, uint, ulong*, const(TypeInfo)) (this=..., size=72, bits=2, alloc_size=0x7fffffffdde8, ti=<optimized out>) at ../../../../src/libphobos/libdruntime/gc/gc.d:489
#6  0x000055555559072d in _d_newclass (ci=0x5555557f8340 <ClassInfo for Exception>) at ../../../../src/libphobos/libdruntime/rt/lifetime.d:96
#7  0x000055555557e209 in std.exception.bailOut!(Exception).bailOut(immutable(char)[], ulong, const(char[])) (file=..., line=line@entry=2496, msg=...)
    at /build/gcc-7-ipontB/gcc-7-7.2.0/src/libphobos/src/std/exception.d:401
#8  0x000055555557e2b1 in std.exception.enforce!(Exception, bool).enforce(bool, lazy const(char)[], immutable(char)[], ulong) (value=<optimized out>, msg=..., file=..., 
    line=line@entry=2496) at /build/gcc-7-ipontB/gcc-7-7.2.0/src/libphobos/src/std/exception.d:363
#9  0x000055555557bf3c in std.stdio.File.LockingTextWriter.this(ref std.stdio.File) (f=..., this=<synthetic pointer>...) at ../../../../src/libphobos/src/std/stdio.d:2496
#10 std.stdio.File.lockingTextWriter() (this=...) at ../../../../src/libphobos/src/std/stdio.d:2657
#11 0x00005555555695a5 in std.stdio.File.writefln!(char, int).writefln(const(char[]), int) (this=..., fmt=..., _param_1=1) at /usr/lib/gcc/x86_64-linux-gnu/7/include/d/std/stdio.d:1384
#12 0x00005555555694ee in std.stdio.writefln!(immutable(char)[], int).writefln(immutable(char)[], int) (_param_0=..., _param_1=1)
    at /usr/lib/gcc/x86_64-linux-gnu/7/include/d/std/stdio.d:3289
#13 0x000055555556896a in print_hello(int) (i=1) at ../test cases/d/10 d cpp/libfile.d:4
#14 0x0000555555568932 in main () at ../test cases/d/10 d cpp/cppmain.cpp:4
```

And with this in dcpp:

```
Starting program: /home/jpakkane/workspace/meson/abc/dcpp 
/home/jpakkane/workspace/meson/abc/dcpp: Symbol `_ZNSt8ios_base4InitC1Ev' causes overflow in R_X86_64_PC32 relocation
/home/jpakkane/workspace/meson/abc/dcpp: Symbol `__cxa_atexit' causes overflow in R_X86_64_PC32 relocation
/home/jpakkane/workspace/meson/abc/dcpp: Symbol `_ZStlsISt11char_traitsIcEERSt13basic_ostreamIcT_ES5_PKc' causes overflow in R_X86_64_PC32 relocation
/home/jpakkane/workspace/meson/abc/dcpp: Symbol `_ZStlsISt11char_traitsIcEERSt13basic_ostreamIcT_ES5_PKc' causes overflow in R_X86_64_PC32 relocation
/home/jpakkane/workspace/meson/abc/dcpp: Symbol `_ZNSolsEi' causes overflow in R_X86_64_PC32 relocation
[Thread debugging using libthread_db enabled]
Using host libthread_db library "/lib/x86_64-linux-gnu/libthread_db.so.1".

Program received signal SIGSEGV, Segmentation fault.
0x00005554f7aef3d0 in ?? ()
(gdb) bt
#0  0x00005554f7aef3d0 in ?? ()
#1  0x0000555555564767 in __cxx_global_var_init () at /usr/bin/../lib/gcc/x86_64-linux-gnu/7.2.0/../../../../include/c++/7.2.0/iostream:74
#2  0x00005555555647a9 in _GLOBAL__sub_I_libfile.cpp ()
#3  0x00005555555a9add in __libc_csu_init ()
#4  0x00007ffff7479150 in __libc_start_main (main=0x5555555648ba <main>, argc=1, argv=0x7fffffffe0b8, init=0x5555555a9a90 <__libc_csu_init>, fini=<optimized out>, 
    rtld_fini=<optimized out>, stack_end=0x7fffffffe0a8) at ../csu/libc-start.c:264
#5  0x00005555555647da in _start ()
```

I'm going to make it skip the tests with Clang for now to get this merged. Patches welcome to fix it properly.

**Comment by ximion** [comment]
Would be interesting to know if this is different when compiling the D bits with LDC and the C++ bits with Clang (so both compilers use the same backend).
But yeah, skipping the tests for Clang sounds like a good idea to me.

**Pull request merged**